### ISSUE TITLE
Using toolcache on hosted pools

### ISSUE THREAD

**Issue by danechitoaie** [actions/toolkit#227]
I have an action with the code bellow and each time I run the action it goes on the ` if (!toolDir) {` branch since the `toolCache.find` never returns anything.
So looks like my file is never cached.

```
        let toolVer = "1.0.0";
        let toolDir = toolCache.find("tool-name", toolVer);

        if (!toolDir) {
            const tempDirectory = process.env.RUNNER_TEMP;
            if (!tempDirectory) {
                tempDirectory = path.join("/home", "actions", "temp");
            }

            const destDirectory = path.join(tempDirectory, uuidV4());
            await io.mkdirP(destDirectory);

            // some code to download a file in the above destDirectory

            toolDir = await toolCache.cacheDir(destDirectory, "tool-name", toolVer);
        }
```

**Comment by thboop** [comment]
Hey @danechitoaie ,

Are you using a self hosted runner? The hosted runners reset the machines to a known image in between jobs, you won't be able to cache things in the tool cache on them. The tool cache is intended for self hosted runners, or in the case where you have to switch back and forth between versions of a tool multiple times in one job.

If you are using a self hosted runner, you don't want to use the RUNNER_TEMP directory, we clean that out in between each job.

**Comment by danechitoaie** [comment]
Hi. I was using the hosted runner. I was putting something in the temp folder just temporarily (as it's downloaded from Azure BlobStorage) before using `toolCache.cacheDir`.

Ok, so it doesn't work with hosted workers. Clear now. Maybe documentation can be updated to reflect this.

Thanks.

**Comment by bryanmacfarlane** [comment]
note that toolcache and the cache dir is all about caching downloaded tools (node, go, etc.) by platform, version.  So it does work on hosted - it's just that you may not get the perf win on hosted.  How are you using the tool cache and what are you downloading from blob storage?  Also, are you depending on it or is a pure cache and perf win?

**Comment by danechitoaie** [comment]
I have a custom GitHub action that's a wrapper on top of a Java fatjar CLI app. Since GitHub does not support private GitHub Actions I opted in making my action public but the action downloads the jar from Azure Blob Storage  and runs in instead of having it bundled with the action. So I was hoping I would be able to cache the jar for a bit and not download it all the time for all the builds from BlobStorage.
That's kind of a workaround until GitHub will hopefully support private actions as well.

**Comment by bryanmacfarlane** [comment]
Yeah, hosted is a fresh VM every time so anything you download is gone the next time.  On self hosted, you will benefit from it being there.  It will work but you just won't benefit from the perf from run to run.

Even with private actions, the bits / action will still need to be download so not exactly sure why private actions would help here.

**Comment by danechitoaie** [comment]
> not exactly sure why private actions would help here.
Private actions would help since if the action is private then I can keep all of my code there and the job downloads only the action (with everything in it).

Now, because we can't have private actions, and because I can put our code in a public repo (action) I have this "workaround". Public repo (action) that's just a wrapper on the underlying app that's downloaded from blob storage. So now two steps/downloads are happening: the normal action being downloaded by GitHub actions and then my private code by the action before being able to execute it.

Basically private actions would allow me to bundle my sensitive/private data/code along with it.
